node/src/components/consensus/tests/mock_proto.rs

```diff
@@ -0,0 +1,253 @@
+pub(crate) struct NodeId(pub u8);
+
+/// A mock network message.
+/// These are basically instructions to the `MockProto` to trigger certain results.
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(bound(
+    serialize = "C::ValidatorId: Serialize",
+    deserialize = "C::ValidatorId: Deserialize<'de>",
+))]
+pub(crate) enum Message<C: Context> {
+    /// Evidence that a validator has equivocated.
+    Evidence(C::ValidatorId),
+    /// A block proposed by another validator.
+    BlockByOtherValidator {
+        value: C::ConsensusValue,
+        timestamp: Timestamp,
+        proposer: C::ValidatorId,
+    },
+    /// Represents network messages that cause the next block to be finalized.
+    FinalizeBlock,
+}
+
+impl<C: Context> Message<C>
+where
+    C::ValidatorId: Serialize,
+{
+    /// Creates a `MessageReceived` event from this message.
+    pub(crate) fn received(&self, sender: NodeId, era_id: EraId) -> Event<NodeId> {
+        let payload = bincode::serialize(self).expect("should serialize message");
+        let msg = ConsensusMessage::Protocol { era_id, payload };
+        Event::MessageReceived { sender, msg }
+    }
+}
+
+/// A block that was proposed but not finalized yet.
+#[derive(DataSize, Debug)]
+pub(crate) struct PendingBlock<C>
+where
+    C: Context,
+{
+    value: C::ConsensusValue,
+    timestamp: Timestamp,
+    proposer: C::ValidatorId,
+}
+
+/// A mock consensus protocol for testing.
+#[derive(DataSize, Debug)]
+pub(crate) struct MockProto<C>
+where
+    C: Context,
+{
+    /// The unique ID of this protocol instance.
+    instance_id: C::InstanceId,
+    /// Validators that were observed to be faulty.
+    evidence: BTreeSet<C::ValidatorId>,
+    /// Validators marked as faulty due to external evidence.
+    faulty: BTreeSet<C::ValidatorId>,
+    /// This validator's ID and secret key, if active.
+    active_validator: Option<(C::ValidatorId, C::ValidatorSecret)>,
+    pending_blocks: VecDeque<PendingBlock<C>>,
+    finalized_blocks: Vec<FinalizedBlock<C>>,
+}
+
+impl<C: Context + 'static> MockProto<C>
+where
+    C::ValidatorId: Serialize + DeserializeOwned,
+{
+    /// Creates a new boxed `MockProto` instance.
+    pub(crate) fn new_boxed(
+        instance_id: C::InstanceId,
+        _validator_stakes: Vec<(C::ValidatorId, Motes)>,
+        _slashed: &HashSet<C::ValidatorId>,
+        _chainspec: &Chainspec,
+        _prev_cp: Option<&dyn ConsensusProtocol<NodeId, C>>,
+        _start_time: Timestamp,
+        _seed: u64,
+    ) -> Box<dyn ConsensusProtocol<NodeId, C>> {
+        Box::new(MockProto {
+            instance_id,
+            evidence: Default::default(),
+            faulty: Default::default(),
+            active_validator: None,
+            pending_blocks: Default::default(),
+            finalized_blocks: Default::default(),
+        })
+    }
+}
+
+impl<C: Context + 'static> ConsensusProtocol<NodeId, C> for MockProto<C>
```

Comment:
And why we didn't want to reuse `HighwayProtocol` and the production version of `impl ConsensusProtocol for HighwayProtocol`?

Reply:
Then we can't easily test a single instance of the era supervisor in isolation.
`MockProto` allows us to explicitly decide in the test cases when a block should be considered proposed, finalized, when a validator should be seen as faulty, etc.

Reply:
But we could have done the same with `HighwayProtocol`, since it's also a state machine that changes its state (and produces messages) based only on the input messages?

Reply:
Sure, but it's very complicated to setup a test case where a certain equivocation or finalization happens, and it involves lots of Highway-specific details. The only somewhat convenient way to create the required messages is to run multiple instances of the era supervisor, one for each validator.

I'm not necessarily against that approach; but I'm struggling to find out how to write tests for cross-era slashing in that case.

Reply:
It's water under the bridge now :) I am just always wary of solutions that test piece of production logic while not reusing the code/mocking a lot.

Reply:
I agree, but I don't see an alternative here. I'm open to suggestions.